client/auth.go

```diff
@@ -69,35 +77,53 @@ func (c *Conn) readInitialHandshake() error {
 	pos += 2
 
 	if len(data) > pos {
-		// skip server charset
+		// default server a_protocol_character_set, only the lower 8-bits
 		// c.charset = data[pos]
 		pos += 1
 
 		c.status = binary.LittleEndian.Uint16(data[pos : pos+2])
 		pos += 2
-		// capability flags (upper 2 bytes)
+
+		// The upper 2 bytes of the Capabilities Flags
 		c.capability = uint32(binary.LittleEndian.Uint16(data[pos:pos+2]))<<16 | c.capability
 		pos += 2
 
-		// auth_data is end with 0x00, min data length is 13 + 8 = 21
-		// ref to https://dev.mysql.com/doc/internals/en/connection-phase-packets.html#packet-Protocol::Handshake
-		maxAuthDataLen := 21
-		if c.capability&CLIENT_PLUGIN_AUTH != 0 && int(data[pos]) > maxAuthDataLen {
-			maxAuthDataLen = int(data[pos])
+		// length of the combined auth_plugin_data (scramble), if auth_plugin_data_len is > 0
+		authPluginDataLen := data[pos]
+		if (c.capability&CLIENT_PLUGIN_AUTH == 0) && (authPluginDataLen > 0) {
+			return errors.Errorf("invalid auth plugin data filler %d", authPluginDataLen)
 		}
+		pos++
 
-		// skip reserved (all [00])
-		pos += 10 + 1
+		// skip reserved (all [00] ?)
+		pos += 10
 
-		// auth_data is end with 0x00, so we need to trim 0x00
-		resetOfAuthDataEndPos := pos + maxAuthDataLen - 8 - 1
-		c.salt = append(c.salt, data[pos:resetOfAuthDataEndPos]...)
+		if c.capability&CLIENT_SECURE_CONNECTION != 0 {
+			// Rest of the plugin provided data (scramble)
 
-		// skip reset of end pos
-		pos = resetOfAuthDataEndPos + 1
+			// $len=MAX(13, length of auth-plugin-data - 8)
+			rest := int(authPluginDataLen) - 8
+			if max := 13; rest > max {
```

Comment:
```suggestion
			if max := 13; rest < max {
```

we should increase the length to 13 😄